integration/data/test_snapshot.py

```diff
@@ -716,42 +718,45 @@ def filesystem_trim_test(dev, address, volume_name, engine_name, fs_type):  # NO
     subprocess.check_call(mount_cmd)
     subprocess.check_call(findmnt_cmd)
 
+    fs_block_size = 4 * 1024
     subprocess.check_call(nsenter_cmd + ["sync"])
-    info = cmd.snapshot_info(address)
+
+    # For ext4, the filesystem metadata contains at least 2 blocks:
+    #     super block + GDT info, inode bitmap.
+    # For xfs, the filesystem metadata contains multiple blocks (> 2 blocks):
+    #     super block, free space B+ tree, inode B+ tree.
+    info = wait_for_snap_size_no_less_than_value(
+        address, VOLUME_HEAD, 2*fs_block_size)
     assert len(info) == 1
-    assert VOLUME_HEAD in info
     fs_meta_size = int(info[VOLUME_HEAD]["size"])
 
-    fs_block_size = 4 * 1024
     length = 3 * 1024
     # create snapshot1 with fs and file 1
     file_path1 = mnt_path + "/test1"
     test1_checksum = write_filesystem_file(length, file_path1)
     subprocess.check_call(nsenter_cmd + ["sync"])
     snap1 = cmd.snapshot_create(address)
 
+    # it is hard to predict the exact blocks the filesystem would consume
+    # when writing a file. but writing a 3Ki file requires at least 3 blocks:
+    # the data itself, inode info update, and free space info (for xfs)
+    info = wait_for_snap_size_no_less_than_value(
```

Comment:
Is the snapshot synchronous? If it's, why need an extra wait here? I know you want to assume the size is expected, but if it's not expected, at the end the test case will still fail in a later step. So, adding the wait here seems not really helpful?

Reply:
I am afraid of the data is not completely flushed into the volume head file before snapshotting. But YES you are right, I should check and wait for the volume head file size before snapshotting, rather than do it after snapshotting.

Reply:
Updated the PR. Added the volume head size check before snapshot creation, which help guarantee that the data is flushed into the file before snapshotting.